arbstate/daprovider/reader.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package daprovider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/util/blobs"
+)
+
+type Reader interface {
+	// IsValidHeaderByte returns true if the given headerByte has bits corresponding to the DA provider
+	IsValidHeaderByte(headerByte byte) bool
+
+	// RecoverPayloadFromBatch fetches the underlying payload from the DA provider given the batch header information
+	RecoverPayloadFromBatch(
+		ctx context.Context,
+		batchNum uint64,
+		batchBlockHash common.Hash,
+		sequencerMsg []byte,
+		preimages map[arbutil.PreimageType]map[common.Hash][]byte,
+		keysetValidationMode KeysetValidationMode,
```

Comment:
keysetValidationmessage is another weird param.
Some other options:
* have a separate validateSequencerMessage that does the validation only, have the caller warn / ignore / panic when that fails as it sees fit
* replace with a boolen input (validate or don't validate). Add a boolean return value indicating if validation succeeded or failed, or use specific error for validation failed and test it with errors.is

Reply:
Agreed and implemented a fix